apps/include/app_bio_functions.h

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2015-2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OSSL_APP_BIO_FUNCTIONS_H
+# define OSSL_APP_BIO_FUNCTIONS_H
+
+# include <openssl/bio.h>
+
+/*
+ * These are used by the macros bio_in, bio_out and bio_err, defined in
+ * app_bio_macros.h.  They are intended for use as lvalues just as much
+ * as rvalues.
+ */
+BIO **app_bio_in_location(void);
```

Comment:
Was it really that bad being able to see the BIOS as externs?

Reply:
`errno` is also an lvalue, and yes in the presence of threads.  The three `bio_xxx` commands aren't lvalues:
```
; g grep -E 'bio_(in|out|err) *=' apps 
apps/cmp.c:    bio_err = NULL; /* do not show errors on more than one try */
apps/cmp.c:    bio_err = bio_bak;
apps/cmp.c:        bio_err = NULL; /* suppress diagnostics on first try loading key */
apps/cmp.c:        bio_err = bio_bak;
apps/openssl.c:BIO *bio_in = NULL;
apps/openssl.c:BIO *bio_out = NULL;
apps/openssl.c:BIO *bio_err = NULL;
apps/openssl.c:    bio_in = dup_bio_in(FORMAT_TEXT);
apps/openssl.c:    bio_out = dup_bio_out(FORMAT_TEXT);
apps/openssl.c:    bio_err = dup_bio_err(FORMAT_TEXT);
```

Well, except for cmp of course. :)  There's probably better ways to do what it's doing.

Reply:
> The three `bio_xxx` commands aren't lvalues:

You *do* know that the "l" in "lvalue" is for "left", right?  Your grep output shows `bio_err`, `bio_in` and `bio_out` used, precisely, as lvalues.

Reply:
Yes, thank you I know what `lvalue` means.  My point was that they're really globals initialized in the main apps loop, except for the `cms` command.

The global `= NULL;` initialization isn't necessary, but is (was?) apparently openssl style to handle some broken C runtimes or such.